src/fs.rs

```diff
@@ -143,22 +143,45 @@ impl BiosParameterBlock {
         bpb.total_sectors_32 = rdr.read_u32::<LittleEndian>()?;
 
         // sanity checks
-        if bpb.bytes_per_sector < 512 {
+        if bpb.bytes_per_sector.count_ones() != 1 {
             return Err(Error::new(ErrorKind::Other, "invalid bytes_per_sector value in BPB"));
-        }
-        if bpb.sectors_per_cluster < 1 {
+        } else if bpb.bytes_per_sector < 512 {
+            return Err(Error::new(ErrorKind::Other, "invalid bytes_per_sector value in BPB"));
+        } else if bpb.bytes_per_sector > 4096 {
+            return Err(Error::new(ErrorKind::Other, "invalid bytes_per_sector value in BPB"));
+        } 
+
+        if bpb.sectors_per_cluster.count_ones() != 1 {
+            return Err(Error::new(ErrorKind::Other, "invalid sectors_per_cluster value in BPB"));
+        } else if bpb.sectors_per_cluster < 1 {
+            return Err(Error::new(ErrorKind::Other, "invalid sectors_per_cluster value in BPB"));
+        } else if bpb.sectors_per_cluster > 128 {
             return Err(Error::new(ErrorKind::Other, "invalid sectors_per_cluster value in BPB"));
         }
+
+        // bytes per sector is u16, sectors per cluster is u8, so guaranteed no overflow in multiplication
+        let bytes_per_cluster : u32 = bpb.bytes_per_sector as u32 * bpb.sectors_per_cluster as u32; 
+        let maximum_bytes_per_cluster : u32 = u32::max_value();
+        ////Microsoft documents indicate 32k is maximum for compatibility; 64k ***might*** work on some platforms
+        //let maximum_bytes_per_cluster : u32 = 64 * 1024;
+        //let maximum_bytes_per_cluster : u32 = 32 * 1024;
+        if bytes_per_cluster > maximum_bytes_per_cluster {
+            return Err(Error::new(ErrorKind::Other, "bytes_per_cluster (bytes_per_sector * sectors_per_cluster) too large in BPB"));
+        } else
+
+        // warning: fat12 and fat16 code exists that presume value == 1
```

Comment:
This is a warning because any value other than 1 will cause the file system to be incompatible with at least some implementations of FAT.  I've changed this to use the warn! macro.